### PR TITLE
Srashid/chart legend/#142959855

### PR DESCRIPTION
**Pull request by srashid5**
Please review and let me know if i need to change / fix anything.

### CODE REVIEW

File: app/models/concerns/project_statistics.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  private
+
+  def projects_sum_of_attribute_value(attr)
+    self.projects.reduce(0) {|x, p| x + (p.public_send(attr) || 0)}
+  end
+
+  def projects_count_by_attribute(attr, value)
+    self.projects.select {|x| x if x.public_send(attr) == value.to_s}.count
```
Comment (mckinley) [inline app/models/concerns/project_statistics.rb]: In the code above, all of the projects in a Portfolio or OpportunityArea are loaded, and active record objects are instantiated for each record, and finally filtered and counted by ruby. This is extremely inefficient.

on the other hand the following code would retrive just the count, not instantiate any records, and is much easier to read and write as well:

```
self.projects.where(attr => value).count
```

Its so simple that we could probably remove these private methods and simply call the methods in the public methods:

```
def projects_active
   self.projects.where(status: active).count
end
```

An even better approach (simpler, efficient, and readable), would be to make a scope on project:

```
# in project.rb
scope :active, -> { where(active: true) }
```

Then you could just call @opportunity_area.projects.active.count or @portfolio.projects.active.count anywhere and you could totally remove this file.

This would be my prefered implementation.

File: app/models/opportunity_area.rb [hunk 2]
```diff
@@ -1,15 +1,8 @@
 class OpportunityArea < ApplicationRecord
   include LastUpdateDelegation
+  include ProjectStatistics
   delegate_to_last_update :budget, :spend
   belongs_to :portfolio
   has_many :projects
   has_many :experiments, through: :projects
-
-  def projects_budget_total
-    projects.reduce(0) { |x, p| x + (p.budget || 0) }
```
Comment (mckinley) [inline app/models/opportunity_area.rb]: I realize that you are following this pattern. But there is a much more efficient way to do this. I apologize for leaving this here. There is a `sum` method in ActiveRecord that should be used instead. The only caveat is that `projects.budget` will probably have to be given a default value of 0 in a migration.

File: app/models/concerns/project_statistics.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  def projects_killed
+    projects_count_by_attribute :status, :killed
+  end
+
+  def projects_pre_seed
+    projects_count_by_attribute :stage, :conceptual
+  end
+
+  def projects_high_potential
+    projects_count_by_high_potenial
```
Comment (srashid5) [inline app/models/concerns/project_statistics.rb]: my bad! just fixed it

File: app/models/concerns/project_statistics.rb (hunk 1 above)
Comment (srashid5) [inline app/models/concerns/project_statistics.rb]: I agree on this being inefficient and I'm very much willing to change this implementation as I wanted to implement this through scopes in the first place. But then, there were a few things I'd like to get your suggestions on.

Assuming we add scopes for (active, on_hold, killed, high_potential, revenue_generating etc) in project.rb, this will add class methods in said class, so from the controller, it will look something like

`@project_active = Project.active.count`
`@project_on_hold = Project.on_hold.count`

which will do a query in projects, joining project_updates (only last updated row), and return the count of all rows with active, on_hold status. Is this correct?

if above assumption is correct then for Portfolio and OpportunityArea classes, similar scopes would be added with joins from project, project_updates?

File: app/models/opportunity_area.rb (hunk 2 above)
Comment (srashid5) [inline app/models/opportunity_area.rb]: I'll change this and use scopes as you've mentioned above.

Yes, I agree, ActiveRecord sum for spend/total and count for project statuses/high_potential/revenue etc

File: app/models/concerns/project_statistics.rb (hunk 1 above)
Comment (mckinley) [inline app/models/concerns/project_statistics.rb]: Yes. Just to be clear, because we already have a has_many :through declaration, we will automatically be able to say
```
@portfolio = Portfolio.find(params[:id])
@portfolio.projects.active.count
```
Not sure if thats relevant...

Guess Im not sure what scopes we would need to add to OpportunityAreas and Portfolios. But If you have an idea about how it should work, go for it.

Also we will mostly be using these in JSONAPI resources. Have a look at `#updates_count` in https://github.com/bionic-solution/bionic-dashboard/blob/master/app/resources/portfolio_resource.rb

That might be relevant.

**Comment by mckinley** [comment]
We want to keep all correspondence related to an issue here or on PT. But I forget sometimes! @srashid5 this is our conversation so far. No need to re-read :). Email correspondence: 
```
Bronson McKinley <bmckinley@bionicsolution.com>
5:00 PM (14 hours ago)

to srashid 
I pushed some changes to your branch that will hopefully make working with entities and their updates more straight-forward. Also the relationship is more obvious. In the example of projects, projects and project_updates are no longer kept in sync automatically. instead we now need to call @project_update.sync_project to update a project after creating a new project_update. Hopefully these changes will make all our queries easier and there is less magic which is probably better.

Muhammad Saad Rashid
6:31 AM (58 minutes ago)

to me 
Hey Bronson,

I saw your changes and they look awesome to me in terms of simplicity and readability.

I just made a very small change. 
I've put all the sync methods in a callback to auto update all parent entities on each create. 
After I push these changes, you will need to rake db:reset on your local so you'l have correctly sync'd data on all parent entities (portfolio, opportunity_areas and projects)

Thanks again for simplifying this. 

Bronson McKinley <bmckinley@bionicsolution.com>
7:11 AM (18 minutes ago)

to Muhammad 
You can do that if you want. I actually like the simplicity of just calling `syc` when its needed. Less magic.

To be honest, about 80% of the time I thought I was being clever by using AR callbacks, it was a mistake. But I never learn my lesson! Ha!

Anyway Im fine with it either way. But lets get it up and move forward. 

Thanks!
```

File: app/models/opportunity_area_update.rb [hunk 4]
```diff
@@ -1,4 +1,9 @@
 class OpportunityAreaUpdate < ApplicationRecord
   default_scope { order(:occurred_at) }
   belongs_to :opportunity_area
+  after_create :sync_opportunity_area
```
Comment (srashid5) [inline app/models/opportunity_area_update.rb]: Oh okay, this is interesting to know. 
>  Unfortunately there is also the complication of an opportunity_area being edited directly.

I understand why you prefer this without hooks. I'l just revert this change, and let's move forward.
